File: backend/src/main/java/com/pgvaale/backend/security/CustomUserDetailsService.java
package com.pgvaale.backend.security;

import com.pgvaale.backend.entity.*;
import com.pgvaale.backend.repository.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import java.util.Collections;
import java.util.Optional;

@Service
public class CustomUserDetailsService implements UserDetailsService {
    
    @Autowired
    private UserRepository userRepository;
    
    @Autowired
    private AdminRepository adminRepository;
    
    @Autowired
    private OwnerRepository ownerRepository;
    
    @Autowired
    private MaidRepository maidRepository;
    
    @Autowired
    private TiffinRepository tiffinRepository;

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        // Try to find user in each repository
        Optional<User> user = userRepository.findByUsername(username);
        if (user.isPresent()) {
            return createUserDetails(user.get(), "ROLE_USER");
        }
        
        Optional<Admin> admin = adminRepository.findByUsername(username);
        if (admin.isPresent()) {
            return createUserDetails(admin.get(), "ROLE_ADMIN");
        }
        
        Optional<Owner> owner = ownerRepository.findByUsername(username);
        if (owner.isPresent()) {
            return createUserDetails(owner.get(), "ROLE_OWNER");
        }
        
        Optional<Maid> maid = maidRepository.findByUsername(username);
        if (maid.isPresent()) {
            return createUserDetails(maid.get(), "ROLE_MAID");
        }
        
        Optional<Tiffin> tiffin = tiffinRepository.findByUsername(username);
        if (tiffin.isPresent()) {
            return createUserDetails(tiffin.get(), "ROLE_TIFFIN");
        }
        
        throw new UsernameNotFoundException("User not found: " + username);
    }
    
    private UserDetails createUserDetails(BaseEntity entity, String role) {
        GrantedAuthority authority = new SimpleGrantedAuthority(role);
        return new org.springframework.security.core.userdetails.User(
                entity.getUsername(),
                entity.getPassword(),
                true, // enabled
                true, true, true, // accountNonExpired, credentialsNonExpired, accountNonLocked
                Collections.singletonList(authority)
        );
    }
} 

File: PGVaaleDotNetBackend/Security/CustomUserDetailsService.cs
using Microsoft.AspNetCore.Identity;
using PGVaaleDotNetBackend.Entities;
using PGVaaleDotNetBackend.Repositories;
using System.Security.Claims;

namespace PGVaaleDotNetBackend.Security
{
    public class CustomUserDetailsService : IUserDetailsService
    {
        private readonly IUserRepository _userRepository;
        private readonly IAdminRepository _adminRepository;
        private readonly IOwnerRepository _ownerRepository;
        private readonly IMaidRepository _maidRepository;
        private readonly ITiffinRepository _tiffinRepository;

        public CustomUserDetailsService(
            IUserRepository userRepository,
            IAdminRepository adminRepository,
            IOwnerRepository ownerRepository,
            IMaidRepository maidRepository,
            ITiffinRepository tiffinRepository)
        {
            _userRepository = userRepository;
            _adminRepository = adminRepository;
            _ownerRepository = ownerRepository;
            _maidRepository = maidRepository;
            _tiffinRepository = tiffinRepository;
        }

        public async Task<UserDetails> LoadUserByUsernameAsync(string username)
        {
            // Try to find user in each repository
            var user = await _userRepository.FindByUsernameAsync(username);
            if (user != null)
            {
                return CreateUserDetails(user, "ROLE_USER");
            }

            var admin = await _adminRepository.GetByUsernameAsync(username);
            if (admin != null)
            {
                return CreateUserDetails(admin, "ROLE_ADMIN");
            }

            var owner = await _ownerRepository.FindByUsernameAsync(username);
            if (owner != null)
            {
                return CreateUserDetails(owner, "ROLE_OWNER");
            }

            var maid = await _maidRepository.FindByUsernameAsync(username);
            if (maid != null)
            {
                return CreateUserDetails(maid, "ROLE_MAID");
            }

            var tiffin = await _tiffinRepository.FindByUsernameAsync(username);
            if (tiffin != null)
            {
                return CreateUserDetails(tiffin, "ROLE_TIFFIN");
            }

            throw new UsernameNotFoundException($"User not found: {username}");
        }

        private UserDetails CreateUserDetails(BaseEntity entity, string role)
        {
            var claims = new List<Claim>
            {
                new Claim(ClaimTypes.Name, entity.Username),
                new Claim(ClaimTypes.Role, role),
                new Claim("UserId", entity.Id.ToString())
            };

            var identity = new ClaimsIdentity(claims, "Custom");
            var principal = new ClaimsPrincipal(identity);

            return new CustomUserDetails
            {
                Username = entity.Username,
                Password = entity.Password,
                Role = role,
                UserId = entity.Id,
                IsEnabled = true,
                ClaimsPrincipal = principal
            };
        }
    }

    public class CustomUserDetails : UserDetails
    {
        public override string Username { get; set; } = string.Empty;
        public override string Password { get; set; } = string.Empty;
        public override bool IsEnabled { get; set; }
        public string Role { get; set; } = string.Empty;
        public long UserId { get; set; }
        public ClaimsPrincipal ClaimsPrincipal { get; set; } = null!;
    }

    public class UsernameNotFoundException : Exception
    {
        public UsernameNotFoundException(string message) : base(message)
        {
        }
    }
}
